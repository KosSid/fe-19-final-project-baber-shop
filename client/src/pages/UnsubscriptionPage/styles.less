.unsubscribe {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  min-height: 380px;
  font-size: 2vw;

  & .icon {
    padding-bottom: 50px;
    font-size: 5vw;

    &-good {
      color: forestgreen;
    }

    &-bad {
      color: red;
    }
  }
}