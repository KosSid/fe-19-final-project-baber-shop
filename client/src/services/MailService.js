import Ajax from "./Ajax";

class MailService {
    OK = 'ok';
    NOK = 'nok';

    async subscribe(email) {
        const subscriptionStatus = await this.checkSubscriptionStatus(email);
        if (subscriptionStatus != null) {
            if (!subscriptionStatus) {
                // existing unsubscribed customer
                try {
                    const {checksum, url} = this.prepareUnsubscribeData(email);

                    await this.updateCustomer(email, {
                        enabled: true,
                        letterSubject: 'Welcome back to Barber shop',
                        letterHtml: `<h1>Hello again and welcome back!</h1><div><a href="${url}">Unsubscribe :(</a></div>`,
                        checksum
                    })
                    return {
                        status: this.OK,
                        message: 'Welcome back!'
                    }
                } catch (err) {

                }
            } else {
                // already subscribed
                return {
                    status: this.NOK,
                    message: 'Email is already subscribed.'
                }
            }
        } else {
            // new customer
            try {
                const {checksum, url} = this.prepareUnsubscribeData(email);

                await Ajax.post('/subscribers', {
                    email,
                    letterSubject: 'Welcome to Barber shop',
                    letterHtml: `<h1>Hello and welcome!</h1><div><a href="${url}">Unsubscribe :(</a></div>`,
                    checksum
                });
                return {
                    status: this.OK,
                    message: 'Successfully subscribed.'
                }
            } catch (err) {
                //
            }
        }
        return {
            status: this.NOK,
            message: 'Internal error, please try again later :('
        }
    }

    prepareUnsubscribeData(email) {
        const checksum = btoa(Math.random().toString(36).substring(7));
        const url = window.location.protocol + '//' + window.location.host + `/unsubscribe?email=${encodeURIComponent(email)}&check=${encodeURIComponent(checksum)}`;
        return {checksum, url};
    }

    async unsubscribe(email, checksum) {
        try {
            const subs = await this.getSubscriber(email);

            if (subs.checksum === checksum) {
                await this.updateCustomer(email, {
                    enabled: false,
                    letterSubject: 'Barbershop says goodbye!',
                    letterHtml: '<h1>Thanks and goodbye!</h1>',
                    checksum: this.prepareUnsubscribeData(email).checksum
                })
                return true;
            }
        } catch (err) {
            //
        }
        return false;
    }

    async getSubscriber(email) {
        try {
            return await Ajax.get(`/subscribers/${email}`);
        } catch (err) {
            return null;
        }
    }

    async checkSubscriptionStatus(email) {
        try {
            return (await this.getSubscriber(email)).enabled;
        } catch (err) {
            return null;
        }
    }

    async updateCustomer(email, data) {
        await Ajax.put('/subscribers/email', email, data)
    }
}

export default new MailService();