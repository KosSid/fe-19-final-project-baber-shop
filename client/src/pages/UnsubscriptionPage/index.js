import React, {useEffect, useState} from "react";
import {useLocation} from "react-router-dom";
import queryString from 'query-string';
import MailService from "../../services/MailService";
import {CheckOutlined, MehOutlined} from "@ant-design/icons";
import './styles.less';
import Preloader from "../../components/Preloader";

const UnsubscriptionPage = () => {
    const location = useLocation();
    const [state, setState] = useState(null);

    useEffect(() => {
        const query = queryString.parse(location.search);
        MailService.unsubscribe(query.email, query.check)
            .then(res => setState(res))
    }, [location]);

    return (
        <div className='unsubscribe'>
            {
                state !== null ?
                    state ?
                        <>
                            <CheckOutlined className='icon icon-good'/>
                            All good!
                        </>
                        :
                        <>
                            <MehOutlined className='icon icon-bad'/>
                            Sorry, you went to the wrong place.
                        </>
                    : <Preloader/>
            }
        </div>
    );
}

export default UnsubscriptionPage;