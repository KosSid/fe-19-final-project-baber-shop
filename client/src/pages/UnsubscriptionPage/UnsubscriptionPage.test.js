import React from 'react'
import UnsubscriptionPage from './index'
import MailService from '../../services/MailService'
import {render, unmountComponentAtNode} from 'react-dom'
import {act} from '@testing-library/react'

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
})

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
})

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:3000/example/path',
  }),
}));

test('UnsubscriptionPage is rendered true', async () => {
  jest.spyOn(MailService, 'unsubscribe').mockImplementation(() =>
    Promise.resolve(true)
  );

  await act(async () => {
    await render(<UnsubscriptionPage/>, container);
  });

  expect(container.innerHTML).toMatchSnapshot();

  MailService.unsubscribe.mockRestore();
})

test('UnsubscriptionPage is rendered false', async () => {
  jest.spyOn(MailService, 'unsubscribe').mockImplementation(() =>
    Promise.resolve(false)
  );

  await act(async () => {
    await render(<UnsubscriptionPage/>, container);
  });

  expect(container.innerHTML).toMatchSnapshot();

  MailService.unsubscribe.mockRestore();
})
