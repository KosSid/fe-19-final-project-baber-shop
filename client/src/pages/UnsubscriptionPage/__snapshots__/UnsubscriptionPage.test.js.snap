// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnsubscriptionPage is rendered false 1`] = `"<div class=\\"unsubscribe\\"><span role=\\"img\\" aria-label=\\"meh\\" class=\\"anticon anticon-meh icon icon-bad\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"meh\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M288 421a48 48 0 1096 0 48 48 0 10-96 0zm352 0a48 48 0 1096 0 48 48 0 10-96 0zM512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm263 711c-34.2 34.2-74 61-118.3 79.8C611 874.2 562.3 884 512 884c-50.3 0-99-9.8-144.8-29.2A370.4 370.4 0 01248.9 775c-34.2-34.2-61-74-79.8-118.3C149.8 611 140 562.3 140 512s9.8-99 29.2-144.8A370.4 370.4 0 01249 248.9c34.2-34.2 74-61 118.3-79.8C413 149.8 461.7 140 512 140c50.3 0 99 9.8 144.8 29.2A370.4 370.4 0 01775.1 249c34.2 34.2 61 74 79.8 118.3C874.2 413 884 461.7 884 512s-9.8 99-29.2 144.8A368.89 368.89 0 01775 775zM664 565H360c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h304c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z\\"></path></svg></span>Sorry, you went to the wrong place.</div>"`;

exports[`UnsubscriptionPage is rendered true 1`] = `"<div class=\\"unsubscribe\\"><span role=\\"img\\" aria-label=\\"check\\" class=\\"anticon anticon-check icon icon-good\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"check\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z\\"></path></svg></span>All good!</div>"`;
